devlib/module/sched.py

```diff
@@ -0,0 +1,239 @@
+        return procfs["domain0"].get("group0", {}).get("energy", {}).get("cap_states") != None
+
+    @memoized
+    def _has_dmips_capacity(self, cpu):
+        """
+        :returns: Whether dmips capacity data is available for 'cpu'
+        """
+        return self.target.file_exists(
+            os.path.join(self.sysfs_root, 'cpu/cpu{}/cpu_capacity'.format(cpu))
+        )
+
+    @memoized
+    def _get_em_capacity(self, cpu):
+        """
+        :returns: The maximum capacity value exposed by the EAS energy model
+        """
+        sd = SchedDomainData(self.target, cpu)
+
+        cap_states = sd.domains[0].groups[0].energy.cap_states
+        return int(cap_states.split('\t')[-2])
+
+    @memoized
+    def _get_dmips_capacity(self, cpu):
+        """
+        :returns: The capacity value generated from the capacity-dmips-mhz DT entry
+        """
+        return self.target.read_value(
+            os.path.join(
+                self.sysfs_root,
+                'cpu/cpu{}/cpu_capacity'.format(cpu)
+            ),
+            int
+        )
+
+    @memoized
+    def get_capacities(self):
+        """
+        :returns:  a dictionnary of the shape {cpu : capacity}
+        """
+        cpus = range(0, self.target.number_of_cpus)
+
+        capacities = {cpu : 0 for cpu in cpus}
+
+        for cpu in cpus:
+            if self._has_em(cpu):
+                capacities[cpu] = self._get_em_capacity(cpu)
+            elif self._has_dmips_capacity(cpu):
+                capacities[cpu] = self._get_dmips_capacity(cpu)
+            else:
+                self.logger.warning('Cannot read capacity for cpu{}, assuming 1024'.format(cpu))
+                capacities[cpu] = 1024
+
+        return capacities
+
+    @memoized
+    def get_hz(self):
+        """
+        :returns: The scheduler tick frequency on the target
+        """
+        return int(test_env.target.config.get('CONFIG_HZ'))
+
+class SchedDomainData(object):
+    """
+    An interface to the sched domain data exposed in the procfs
+
+    The raw procfs nodes are available under the 'procfs' attribute.
+    They are also parsed and available as attributes of the object.
+
+    -----
+    Example:
+    data = target.sched.get_sd_info(0)
+
+    print info.domains[0].name
+    > 'MC'
+    print info.procfs["domain0"]["name"]
+    > 'MC'
+    """
+
+    sched_domain_root = '/proc/sys/kernel/sched_domain'
+
+    def _build_domains(self):
+        self.domains = {}
+
+        for domain, fields in self.procfs.iteritems():
+            self.domains[ProcFSNode._node_digits(domain)] = SchedDomain(fields)
+
+    def __init__(self, target, cpu):
+        self.procfs = target.read_tree_values(
+            os.path.join(self.sched_domain_root, 'cpu{}'.format(cpu)),
+            depth=5
+        )
+
+        self._build_domains()
+
+class ProcFSNode(object):
+    re_node = re.compile(r"(?P<name>.*)(?P<digits>\d+)$")
+
+    @staticmethod
+    def _ends_with_digits(node):
```

Comment:
I can't really find an established convention about leading underscore for names in classes, but at module level it means "don't export this in from XXX import *". If that is not part of the "user API", that qualifies for leading underscore even if used by different classes IMHO.

Reply:
The convention is established by PEP8: https://www.python.org/dev/peps/pep-0008/#id47

> Use one leading underscore only for non-public methods and instance variables.

This is to do with hinting the accessibility of class members and nothing to do with imports. `from XXX import *` behavior is controlled by `__all__`: https://docs.python.org/2/tutorial/modules.html#importing-from-a-package

Reply:
I also checked PEP8, non-public could be understood both ways, but it may be clarified in another part of the document I skipped.

PEP8 also adds:
> _single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import > objects whose name starts with an underscore.

Python 2.7.15 and 3.6.5 both behave like that by default, unless __all__ is defined.

Reply:
Fair point, however that refers to top-level definitions (which *can* be imported). For a class member, it unambiguously means "internal to the class".